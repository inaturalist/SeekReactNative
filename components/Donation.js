// @flow

import * as React from "react";
import {
  View,
  TouchableOpacity,
  Image,
  Platform
} from "react-native";
import { WebView } from "react-native-webview";
import { SafeAreaView } from "react-native-safe-area-context";

import i18n from "../i18n";
import { viewStyles, textStyles } from "../styles/species/donation";
import icons from "../assets/icons";
import urls from "../constants/urls";
import CopyButton from "./UIComponents/Buttons/CopyButton";
import StyledText from "./UIComponents/StyledText";

type Props = {
  +navigation: any
};

const Donation = ( { navigation, route }: Props ): React.Node => {
  const goBack = ( ) => navigation.goBack( );

  const [selectedText, setSelectedText] = React.useState( false );

  const { params } = route;

  const standardCampaign = `${urls.DONORBOX}${urls.DONORBOX_STANDARD_CAMPAIGN}&utm_source=android`;
  const seekYearInReviewCampaign = `${urls.DONORBOX}&utm_campaign=${params?.utmCampaign}&utm_source=android`;
  const donationPage = params?.utmCampaign ? seekYearInReviewCampaign : standardCampaign;
  const redirectForiOS = "inaturalist.org/donate-seek";

  const highlightSelectedText = ( ) => setSelectedText( true );

  return (
    <SafeAreaView style={viewStyles.container} edges={["top"]}>
      <View style={viewStyles.header}>
        <StyledText style={textStyles.text}>{i18n.t( "settings.donate" ).toLocaleUpperCase( )}</StyledText>
        <TouchableOpacity
          onPress={goBack}
          style={viewStyles.back}
        >
          <Image source={icons.closeWhite} />
        </TouchableOpacity>
      </View>
      {Platform.OS === "android" ? (
        <WebView
          startInLoadingState
          source={{ uri: donationPage }}
        />
      ) : (
        <View style={viewStyles.whiteContainer}>
          <StyledText style={[textStyles.text, textStyles.blackText]}>
           {i18n.t( "settings.donate_ios" )}
          </StyledText>
          <CopyButton stringToCopy={redirectForiOS} handleHighlight={highlightSelectedText}>
            <StyledText style={[
              textStyles.donateText,
              selectedText && viewStyles.selectedPressableArea
            ]}>
              {redirectForiOS}
            </StyledText>
          </CopyButton>
        </View>
      )}
      <View style={viewStyles.bottom} />
    </SafeAreaView>
  );
};

export default Donation;
