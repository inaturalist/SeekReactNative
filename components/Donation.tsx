import * as React from "react";
import {
  View,
  TouchableOpacity,
  Image,
  Platform
} from "react-native";
import { WebView } from "react-native-webview";
import { SafeAreaView } from "react-native-safe-area-context";

import i18n from "../i18n";
import { viewStyles, textStyles } from "../styles/species/donation";
import icons from "../assets/icons";
import urls from "../constants/urls";
import CopyButton from "./UIComponents/Buttons/CopyButton";
import StyledText from "./UIComponents/StyledText";
import { baseTextStyles } from "../styles/textStyles";

const Donation = ( { navigation, route } ) => {
  const goBack = ( ) => navigation.goBack( );

  const [selectedText, setSelectedText] = React.useState( false );

  const { params } = route;

  const standardCampaign = `${urls.DONORBOX}${urls.DONORBOX_STANDARD_CAMPAIGN}&utm_source=android`;
  const seekYearInReviewCampaign = `${urls.DONORBOX}&utm_campaign=${params?.utmCampaign}&utm_source=android`;
  const donationPage = params?.utmCampaign ? seekYearInReviewCampaign : standardCampaign;
  const redirectForiOS = "inaturalist.org/donate-seek";

  const highlightSelectedText = ( ) => setSelectedText( true );

  return (
    <SafeAreaView style={viewStyles.container} edges={["top"]}>
      <View style={viewStyles.header}>
        <StyledText style={[baseTextStyles.button, textStyles.text]}>
          {i18n.t( "settings.donate" ).toLocaleUpperCase( )}
        </StyledText>
        <TouchableOpacity
          onPress={goBack}
          style={viewStyles.back}
        >
          <Image source={icons.closeWhite} />
        </TouchableOpacity>
      </View>
      {Platform.OS === "android" ? (
        <WebView
          startInLoadingState
          source={{ uri: donationPage }}
        />
      ) : (
        <View style={viewStyles.whiteContainer}>
          <StyledText style={[baseTextStyles.body, textStyles.blackText]}>
           {i18n.t( "settings.donate_ios" )}
          </StyledText>
          <CopyButton stringToCopy={redirectForiOS} handleHighlight={highlightSelectedText}>
            <StyledText style={[
              baseTextStyles.donationLink,
              textStyles.donateText,
              selectedText && viewStyles.selectedPressableArea
            ]}>
              {redirectForiOS}
            </StyledText>
          </CopyButton>
        </View>
      )}
      <View style={viewStyles.bottom} />
    </SafeAreaView>
  );
};

export default Donation;
