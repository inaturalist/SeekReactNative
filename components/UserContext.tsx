import * as React from "react";

const UserContext: Object = React.createContext( null );
const ObservationContext: Object = React.createContext( null );
const SpeciesNearbyContext: Object = React.createContext( null );
const ChallengeContext: Object = React.createContext( null );
const SpeciesDetailContext: Object = React.createContext( null );

export {
  UserContext,
  ObservationContext,
  SpeciesNearbyContext,
  ChallengeContext,
  SpeciesDetailContext
};
