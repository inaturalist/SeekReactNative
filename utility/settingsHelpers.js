// @flow
import Realm from "realm";
import AsyncStorage from "@react-native-async-storage/async-storage";

import realmConfig from "../models/index";
import { setDeviceLanguageOrFallback } from "./languageHelpers";

const getScientificNames = async ( ): Promise<boolean> => {
  try {
    const scientificNames = await AsyncStorage.getItem( "scientific_names" );
    if ( scientificNames === null || scientificNames === "false" ) {
      return false;
    }
    return true;
  } catch ( error ) {
    return false;
  }
};

const toggleLanguage = ( language: string ) => {
  AsyncStorage.setItem( "language", language );
};

const getLanguage = async ( ): Promise<any> => {
  try {
    const language = await AsyncStorage.getItem( "language" );
    return language || setDeviceLanguageOrFallback( );
  } catch ( error ) {
    return false;
  }
};

const getAutoCapture = async ( ): Promise<boolean> => {
  try {
    const camera = await AsyncStorage.getItem( "camera" );
    if ( camera === null || camera === "false" ) {
      return false;
    }
    return true;
  } catch ( error ) {
    return false;
  }
};

const getSeasonality = async ( ): Promise<boolean> => {
  try {
    const seasonality = await AsyncStorage.getItem( "seasonality" );
    if ( seasonality === null || seasonality === "false" ) {
      return false;
    }
    return true;
  } catch ( error ) {
    return false;
  }
};

const setupUserSettings = async ( ) => {
  const realm = await Realm.open( realmConfig );

  try {
    // still keeping these stored in async storage
    // but they'll only be used once to set up this realm
    const autoCapture = await getAutoCapture( );
    const localSeasonality = await getSeasonality( );
    const scientificNames = await getScientificNames( );

    realm.write( ( ) => {
      if ( realm.objects( "UserSettingsRealm" ).length > 0 ) {
        // don't overwrite user settings during setup
        return;
      }
      realm.create( "UserSettingsRealm", {
        autoCapture,
        localSeasonality,
        scientificNames
      }, true );
    } );
  } catch ( e ) {
    console.log( e, "couldn't set up User Settings Realm" );
  }
};

const updateUserSetting = async ( key: string, value: boolean ): Promise<?boolean> => {
  const realm = await Realm.open( realmConfig );
  const userSettings = realm.objects( "UserSettingsRealm" );

  try {
    realm.write( ( ) => {
      userSettings[0][key] = value;
    } );
    return value;
  } catch ( e ) {
    console.log( e, "couldn't update User Settings Realm" );
  }
};

const fetchFromAsyncStorage = async ( key: string ): Promise<?string> => {
  try {
    return await AsyncStorage.getItem( key );
  } catch ( error ) {
    return "";
  }
};

const deleteFromAsyncStorage = async ( key: string ): Promise<boolean> => {
  try {
    await AsyncStorage.removeItem( key );
    return true;
  } catch ( e ) {
    return false;
  }
};

export {
  getScientificNames,
  toggleLanguage,
  getLanguage,
  getAutoCapture,
  getSeasonality,
  setupUserSettings,
  updateUserSetting,
  deleteFromAsyncStorage,
  fetchFromAsyncStorage
};
