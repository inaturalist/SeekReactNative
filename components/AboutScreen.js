// @flow

import React, { useContext } from "react";
import {
  Image,
  View,
  TouchableOpacity
} from "react-native";
import { getVersion, getBuildNumber } from "react-native-device-info";
import { useNavigation } from "@react-navigation/native";
import type { Node } from "react";

import { viewStyles, imageStyles, textStyles } from "../styles/about";
import logos from "../assets/logos";
import i18n from "../i18n";
import { AppOrientationContext, UserContext } from "./UserContext";
import ScrollWithHeader from "./UIComponents/Screens/ScrollWithHeader";
import PrivacyAndTerms from "./UIComponents/PrivacyAndTerms";
import StyledText from "./UIComponents/StyledText";

const AboutScreen = (): Node => {
  const navigation = useNavigation();
  const appVersion = getVersion();
  const buildVersion = getBuildNumber();
  const { login } = useContext( UserContext );
  const { isTablet } = useContext( AppOrientationContext );

  const navToDebug = () => navigation.navigate( "DebugEmailScreen" );
  const disabled = !login;

  return (
    <ScrollWithHeader header="about.header" footer>
      <View style={[viewStyles.textContainer, isTablet && viewStyles.tabletContainer]}>
        <Image source={logos.iNat} />
        <View style={viewStyles.marginSmall} />
        <StyledText style={[textStyles.text, textStyles.boldText]}>{i18n.t( "about.seek" )}</StyledText>
        <StyledText style={textStyles.text}>{i18n.t( "about.joint_initiative" )}</StyledText>
        <View style={viewStyles.block} />
        <Image source={logos.casNatGeo} style={imageStyles.image} />
        <View style={viewStyles.marginLarge} />
        <StyledText style={textStyles.text}>{i18n.t( "about.original" )}</StyledText>
        <View style={viewStyles.margin} />
        <Image source={logos.wwfop} style={imageStyles.wwfop} />
        <View style={viewStyles.marginSmall} />
        <Image source={logos.hhmi} />
        <View style={viewStyles.margin} />
        <StyledText style={[textStyles.text, textStyles.boldText]}>{i18n.t( "about.designed_by" )}</StyledText>
        <StyledText style={textStyles.text}>{i18n.t( "about.inat_team_credits_3" )}</StyledText>
        <View style={viewStyles.marginSmallest} />
        <StyledText style={textStyles.text}>{i18n.t( "about.translators" )}</StyledText>
        <View style={viewStyles.marginSmallest} />
        <StyledText style={textStyles.text}>{i18n.t( "about.join_crowdin" )}</StyledText>
        <PrivacyAndTerms login={login} />
        <TouchableOpacity
          onPress={navToDebug}
          style={viewStyles.debug}
          disabled={disabled}
          testID="debug"
        >
          <StyledText style={textStyles.greenText}>
            {i18n.t( "about.version" ).toLocaleUpperCase()}
            {` ${appVersion} (${buildVersion})`}
          </StyledText>
        </TouchableOpacity>
        <StyledText style={textStyles.text}>
          {i18n.t( "about.help" )}
        </StyledText>
        <View style={viewStyles.block} />
      </View>
    </ScrollWithHeader>
  );
};

export default AboutScreen;
