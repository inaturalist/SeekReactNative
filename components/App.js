// @flow

import React, { useState, useEffect } from "react";
import * as RNLocalize from "react-native-localize";
import {
  setJSExceptionHandler,
  setNativeExceptionHandler
} from "react-native-exception-handler";
import type { Node } from "react";

import RootStack from "./Navigation/RootStack";
import { handleLocalizationChange, loadUserLanguagePreference } from "../utility/languageHelpers";
import { hideLogs } from "../utility/helpers";
import { LanguageContext } from "./UserContext";
import { getLanguage } from "../utility/settingsHelpers";
import SpeciesNearbyProvider from "./Providers/SpeciesNearbyProvider";
import UserLoginProvider from "./Providers/UserLoginProvider";
import ObservationProvider from "./Providers/ObservationProvider";
import AppOrientationProvider from "./Providers/AppOrientationProvider";
import ChallengeProvider from "./Providers/ChallengeProvider";
import SpeciesDetailProvider from "./Providers/SpeciesDetailProvider";
import { log } from "../react-native-logs.config";

const logger = log.extend( "App.js" );

const jsErrorHandler = ( e, isFatal ) => {
  // not 100% sure why jsErrorHandler logs e.name and e.message as undefined sometimes,
  // but I believe it relates to this issue, which reports an unnecessary console.error
  // under the hood: https://github.com/a7ul/react-native-exception-handler/issues/143

  // possibly also related to error boundaries in React 16+:
  // https://github.com/a7ul/react-native-exception-handler/issues/60
  if ( !e.name && !e.message ) {return;}
  logger.error( `JS Error: ${isFatal ? "Fatal:" : ""} ${e.stack}` );
};

// record JS exceptions
setJSExceptionHandler( jsErrorHandler );

// record native exceptions
// only works in bundled mode; will show red screen in dev mode
// tested this by raising an exception in RNGestureHandler.m
// https://stackoverflow.com/questions/63270492/how-to-raise-native-error-in-react-native-app
setNativeExceptionHandler( exceptionString => {
  logger.error( `Native Error: ${exceptionString}` );
} );

const App = ( ): Node => {
  useEffect( () => {
    logger.info( "App start" );
  }, [] );

  const [preferredLanguage, setLanguage] = useState( null );

  const getLanguagePreference = async ( ) => setLanguage( await getLanguage( ) );
  const toggleLanguagePreference = ( ) => getLanguagePreference( );

  const languageValue = { preferredLanguage, toggleLanguagePreference };

  useEffect( ( ) => {
    // wait until check for stored language is completed
    if ( !preferredLanguage ) { return; }
    loadUserLanguagePreference( preferredLanguage );
  }, [preferredLanguage] );

  useEffect( ( ) => {
    hideLogs( );

    // Context
    getLanguagePreference( );

    RNLocalize.addEventListener( "change", handleLocalizationChange );

    return ( ) => RNLocalize.removeEventListener( "change", handleLocalizationChange );
  }, [] );

  return (
    <AppOrientationProvider>
      <UserLoginProvider>
        <LanguageContext.Provider value={languageValue}>
          <ObservationProvider>
            <SpeciesNearbyProvider>
              <ChallengeProvider>
                <SpeciesDetailProvider>
                  <RootStack />
                </SpeciesDetailProvider>
              </ChallengeProvider>
            </SpeciesNearbyProvider>
          </ObservationProvider>
        </LanguageContext.Provider>
      </UserLoginProvider>
    </AppOrientationProvider>
  );
};

export default App;
