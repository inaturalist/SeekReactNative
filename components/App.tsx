import React, { useEffect } from "react";
import {
  setJSExceptionHandler,
  setNativeExceptionHandler
} from "react-native-exception-handler";
import {
  getVersion,
  getBuildNumber
} from "react-native-device-info";
import { SafeAreaProvider } from "react-native-safe-area-context";
import { GestureHandlerRootView } from "react-native-gesture-handler";

import RootStack from "./Navigation/RootStack";
import { hideLogs } from "../utility/helpers";
import { LanguageProvider } from "./Providers/LanguageProvider";
import { SpeciesNearbyProvider } from "./Providers/SpeciesNearbyProvider";
import UserLoginProvider from "./Providers/UserLoginProvider";
import { ObservationProvider } from "./Providers/ObservationProvider";
import { AppOrientationProvider } from "./Providers/AppOrientationProvider";
import { ChallengeProvider } from "./Providers/ChallengeProvider";
import { SpeciesDetailProvider } from "./Providers/SpeciesDetailProvider";
import { log } from "../react-native-logs.config";
import { LogLevels, logToApi } from "../utility/apiCalls";
import ErrorBoundary from "./ErrorBoundary";

const logger = log.extend( "App.tsx" );

const jsErrorHandler = ( e, isFatal ) => {
  // not 100% sure why jsErrorHandler logs e.name and e.message as undefined sometimes,
  // but I believe it relates to this issue, which reports an unnecessary console.error
  // under the hood: https://github.com/a7ul/react-native-exception-handler/issues/143

  // possibly also related to error boundaries in React 16+:
  // https://github.com/a7ul/react-native-exception-handler/issues/60
  // if ( !e.name && !e.message ) {return;}
    if ( isFatal ) {
      logger.error( "Fatal JS Error: ", e.stack );
      logToApi( {
        level: LogLevels.ERROR,
        context: "App.tsx",
        message: e.message,
        errorType: e.constructor?.name,
        backtrace: e.stack
      } );
    } else {
      // This should get logged by ErrorBoundary. For some reason this handler
      // gets called too, so we don't want to double-report errors
    }

};

// record JS exceptions
setJSExceptionHandler( jsErrorHandler, true );

// record native exceptions
// only works in bundled mode; will show red screen in dev mode
// tested this by raising an exception in RNGestureHandler.m
// https://stackoverflow.com/questions/63270492/how-to-raise-native-error-in-react-native-app
setNativeExceptionHandler(
  async ( exceptionString ) => {
    try {
      logToApi( {
        level: LogLevels.ERROR,
        message: exceptionString,
        context: "App.tsx",
        errorType: "native"
      } );
      logger.error( `Native Error: ${exceptionString}` );
    } catch ( e ) {
      // Last-ditch attempt to log something
      logger.error(
        `Native Error: ${exceptionString} (failed to save context)`,
        e
      );
    }
  },
  true, // Force quit the app to prevent zombie states
  true // Enable on iOS
);

const style = { flex: 1 } as const;

const App = ( ) => {
  useEffect( () => {
    hideLogs( );
    logger.info( `App start. Version: ${getVersion()} Build: ${getBuildNumber()}` );
  }, [] );

  return (
    <AppOrientationProvider>
      <UserLoginProvider>
        <LanguageProvider>
          <ObservationProvider>
            <SpeciesNearbyProvider>
              <ChallengeProvider>
                <SpeciesDetailProvider>
                  <GestureHandlerRootView style={style}>
                    <SafeAreaProvider>
                      <ErrorBoundary>
                        <RootStack />
                      </ErrorBoundary>
                    </SafeAreaProvider>
                  </GestureHandlerRootView>
                </SpeciesDetailProvider>
              </ChallengeProvider>
            </SpeciesNearbyProvider>
          </ObservationProvider>
        </LanguageProvider>
      </UserLoginProvider>
    </AppOrientationProvider>
  );
};

export default App;
