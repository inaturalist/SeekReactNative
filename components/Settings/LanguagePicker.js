// @flow

import React, { useContext, useCallback, useMemo, useState } from "react";
import { View, Alert, Platform } from "react-native";
import Checkbox from "react-native-check-box";
import * as RNLocalize from "react-native-localize";
import RNPickerSelect from "react-native-picker-select";
import type { Node } from "react";

import i18n from "../../i18n";
import { viewStyles, textStyles } from "../../styles/settings";
import { colors } from "../../styles/global";
import languages from "../../utility/dictionaries/languageDict";
import { LanguageContext } from "../UserContext";
import { toggleLanguage } from "../../utility/settingsHelpers";
import { deviceLanguageSupported, setDisplayLanguage } from "../../utility/languageHelpers";
import StyledText from "../UIComponents/StyledText";

const localeList = Object.keys( languages ).map( ( locale ) => (
  { value: locale, label: languages[locale].toLocaleUpperCase() }
) );

const placeholder = {};
const pickerStyles = { ...viewStyles };
const showIcon = () => <></>;

const { languageCode } = RNLocalize.getLocales()[0];

const LanguagePicker = (): Node => {
  const { toggleLanguagePreference, preferredLanguage } = useContext( LanguageContext );

  const displayLanguage = setDisplayLanguage( preferredLanguage );
  const isChecked = preferredLanguage === "device" || displayLanguage === languageCode;

  const [pickerValue, setPickerValue] = useState( displayLanguage );

  const handleValueChange = useCallback( ( value ) => {
    // this prevents the double render on new Android install
    // without this, the user changes the language
    // and handleValueChange is immediately called with "en"
    if ( value === displayLanguage && preferredLanguage === "device" ) {
      return;
    }

    // only update state if new language is desired
    if ( value === preferredLanguage ) {
      return;
    }

    // if the user selects language to be set to device language don't show alert
    if ( value === "device" ) {
      // this changes translations on Settings screen in real-time
      i18n.locale = value;
      toggleLanguage( value );
      toggleLanguagePreference();
      return;
    }
    Platform.OS === "ios" ? setPickerValue( value ) : showAlert( value );
  }, [displayLanguage, preferredLanguage, toggleLanguagePreference, showAlert] );

  const showAlert = useCallback( ( value ) => {
    const valueLabel = languages[value];
    Alert.alert( null, i18n.t( "settings.change_language", { language: valueLabel } ), [
      {
        text: i18n.t( "delete.no" ),
        onPress: ( ) => null,
        style: "cancel"
      }, {
        text: i18n.t( "settings.confirm" ),
        onPress: ( ) => {
          // this changes translations on Settings screen in real-time
          i18n.locale = value;
          toggleLanguage( value );
          toggleLanguagePreference();
        }
      }
    ] );
  }, [toggleLanguagePreference] );

  const onDonePress = useCallback( ( ) => {
    showAlert( pickerValue );
  }, [showAlert, pickerValue] );

  const setDeviceLanguage = useCallback( () => handleValueChange( "device" ), [handleValueChange] );

  const renderDeviceCheckbox = useMemo( () => (
    <View style={[viewStyles.row, viewStyles.checkboxRow]}>
      {/* accessibility isn't available for this component, and it's also not
      implemented on iOS for the official react-native-checkbox library
      https://github.com/crazycodeboy/react-native-check-box/issues/94 */}
      <Checkbox
        checkBoxColor={colors.checkboxColor}
        isChecked={isChecked}
        disabled={isChecked}
        onClick={setDeviceLanguage}
        style={viewStyles.checkBox}
      />
      <StyledText style={[textStyles.text, viewStyles.padding]}>{i18n.t( "settings.device_settings" )}</StyledText>
    </View>
  ), [isChecked, setDeviceLanguage] );

  return (
    <View style={viewStyles.donateMarginBottom}>
      <StyledText style={textStyles.header}>{i18n.t( "settings.language" ).toLocaleUpperCase()}</StyledText>
      {deviceLanguageSupported( ) && renderDeviceCheckbox}
      <RNPickerSelect
        hideIcon
        Icon={showIcon}
        items={localeList}
        onValueChange={handleValueChange}
        onDonePress={onDonePress}
        placeholder={placeholder}
        useNativeAndroidPickerStyle={false}
        value={Platform.OS === "ios" ? pickerValue : displayLanguage}
        touchableWrapperProps={{ testID: "picker" }}
        disabled={!displayLanguage}
        style={pickerStyles}
      />
    </View>
  );
};

export default LanguagePicker;
