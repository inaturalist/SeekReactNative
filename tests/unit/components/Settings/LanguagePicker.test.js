import React from "react";
import { render, screen, fireEvent } from "@testing-library/react-native";
import { Alert } from "react-native";

import LanguagePicker from "../../../../components/Settings/LanguagePicker";
import { LanguageContext } from "../../../../components/UserContext";
import { toggleLanguage } from "../../../../utility/settingsHelpers";

// Mock the hooks
jest.mock( "../../../../utility/settingsHelpers" );

const renderPicker = () => {
  render(
    <LanguageContext.Provider
      value={{
        preferredLanguage: "en",
        toggleLanguagePreference: jest.fn()
      }}
    >
      <LanguagePicker />
    </LanguageContext.Provider>
  );
};

const pickerID = "picker";
const newLanguage = "es";
describe( "LanguagePicker", () => {
  test( "should render correctly", async () => {
    renderPicker();
    await screen.findByTestId( pickerID );
    const picker = screen.getByTestId( pickerID );
    expect( picker ).toBeTruthy();
    expect( screen.getByText( "Use device language settings" ) ).toBeTruthy();
    expect( screen ).toMatchSnapshot();
  } );

  test( "should on iOS open alert on Done pressed but not before", async () => {
    const alertSpy = jest.spyOn( Alert, "alert" );
    renderPicker();
    await screen.findByTestId( pickerID );
    const picker = screen.getByTestId( pickerID );

    // Change language to "es" no alert should be shown
    fireEvent( picker, "onValueChange", newLanguage );
    expect( alertSpy ).not.toHaveBeenCalled();
    // Press done, show confirmation alert
    fireEvent( picker, "onDonePress" );
    expect( alertSpy ).toHaveBeenCalledTimes( 1 );
    // TODO: this works with the iOS picker, but not with the Android one
    // because Platform.OS always returns "ios" in the test environment
  } );

  test( "should call the language change hook with the new language", async () => {
    renderPicker();
    await screen.findByTestId( pickerID );
    const picker = screen.getByTestId( pickerID );

    // Change language to "es" no alert should be shown
    fireEvent( picker, "onValueChange", newLanguage );
    fireEvent( picker, "onDonePress" );

    const alertSpy = jest.spyOn( Alert, "alert" );
    // This is a press of the Confirm button on the Alert
    alertSpy.mock.calls[0][2][1].onPress();

    // Expect hook to be called with the new language
    expect( jest.isMockFunction( toggleLanguage ) ).toBeTruthy();
    expect( toggleLanguage ).toBeCalledWith( newLanguage );
  } );

} );
