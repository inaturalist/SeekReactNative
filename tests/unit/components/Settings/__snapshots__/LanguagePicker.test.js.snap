// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguagePicker should render correctly 1`] = `
<View
  style={
    {
      "paddingTop": 16,
    }
  }
>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "#38976d",
          "fontFamily": "Whitney-Semibold",
          "fontSize": 19,
          "letterSpacing": 1.12,
        },
      ]
    }
  >
    IN-APP LANGUAGE
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
        },
        {
          "marginTop": 17,
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "paddingRight": 10.3,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../react-native-check-box/img/ic_check_box.png",
            }
          }
          style={
            {
              "tintColor": "#979797",
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          [
            {
              "color": "#000000",
              "fontFamily": "Whitney-Book",
              "fontSize": 16,
              "lineHeight": 21,
            },
            {
              "paddingTop": 5,
            },
          ],
        ]
      }
    >
      Use device language settings
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignSelf": "stretch",
        },
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="picker"
    >
      <View
        pointerEvents="box-only"
        style={
          {
            "alignItems": "center",
            "paddingVertical": 19,
          }
        }
      >
        <TextInput
          editable={false}
          style={
            [
              {
                "backgroundColor": "#38976d",
                "borderRadius": 6,
                "color": "#ffffff",
                "fontFamily": "Whitney-Semibold",
                "fontSize": 20,
                "letterSpacing": 1.11,
                "paddingBottom": 11,
                "paddingHorizontal": 18,
                "paddingTop": 12,
                "textAlign": "center",
              },
              {},
            ]
          }
          testID="text_input"
          value="ENGLISH"
        />
        <View
          style={
            [
              {
                "position": "absolute",
                "right": 0,
              },
              undefined,
            ]
          }
          testID="icon_container"
        />
      </View>
    </View>
    <Modal
      hardwareAccelerated={false}
      onOrientationChange={[Function]}
      supportedOrientations={
        [
          "portrait",
          "landscape",
        ]
      }
      testID="ios_modal"
      transparent={true}
      visible={false}
    />
  </View>
</View>
`;
